### ISSUE TITLE
createObject does not set cacheControl

### ISSUE THREAD

**Issue by horizon0708** [supabase/storage-api#18]
# Bug report

## Describe the bug

TL;DR When I upload an object to storage via storage-js, the API doesn't care about the cache-control value.

At the moment, when you call `upload()` from the Supabase client, it sets the default cache-control field to 3600 (https://github.com/supabase/storage-js/blob/main/src/lib/StorageFileApi.ts#L15). The API should take this field and set that as max-age (https://github.com/supabase/storage-api/blob/master/src/routes/object/createObject.ts#L59). If this value is not present, it is set to no-cache.

When I upload an object to a bucket, I can see that cache-control is set in the request (see screenshot below). This should mean that the server should respond with `cache-control` header value set to`max-age=3600` when I request for the object later on.
![image](https://user-images.githubusercontent.com/22534929/117560210-cf79e580-b0df-11eb-8e58-33b48a1c4c22.png)

However, when I download the object afterward, the header is set to no-cache.
![image](https://user-images.githubusercontent.com/22534929/117560307-cb9a9300-b0e0-11eb-88da-ac1a3262806c.png)

This is not desirable behavior, as we want to cache as many things as possible for a good user experience (and probably also lessen the load for Supabase too). 

## To Reproduce

1. Using the supabase client, upload any object, with or without an optional `cacheControl` value set.
2. Download the object using the supabase client, refresh the page after the download finishes.
3. on refresh, check the network tab to see that (1) the browser requested and downloaded the object again (2) and the response has `no-cache` header set.

## Expected behavior

Supabase API sets cache-control headers that we pass in.

## System information

- OS: Windows
- Browser (if applies): chrome
- Version of supabase-js:  "1.7.7",
- Version of Node.js: v12.0.0

**Comment by horizon0708** [comment]
The way I'm getting around this atm is to implement/use caching on your app. I'm using [react query](https://react-query.tanstack.com/). I have something like:

```ts
export const fetchAvatar: QueryFunction<string | undefined> = async ({ queryKey }) => {
    const [_key, path] = queryKey
    const { data, error } = await supabaseClient.storage
        .from(AVATAR_BUCKET)
        .download(path as string)
    
    if(error) {
        throw new Error(error.message)
    }

    if(!data) {
        return undefined
    }

    return URL.createObjectURL(data)
}
```
then I pass the query function to react-query with options (note staleTime). 
```ts
export const UserAvartar:React.FC<UserAvatarProps> = ({ name, path, className }) => {
    const { data } = useQuery(
        ['avatar', path], 
        fetchAvatar, 
        { 
            enabled: !!path,
            staleTime: 1000 * 60 * 60, // 1 hour
        }
    )
    return <Avatar className={className} alt={name} src={data} />
}
```
This means that the fetch for an object is cached for an hour (as long as the user doesn't reload the app), so the user won't have to redownload the same image over and over again.

**Comment by horizon0708** [comment]
Checked after updating to 1.11.3. Still happening.

**Comment by horizon0708** [comment]
Did some digging just now. The issue is because the controller is calling `.file()`
```ts
const data = await request.file()
```
This calls [this part of fastify](https://github.com/fastify/fastify-multipart/blob/d6159731892a694b8f91c81ea12ca76f41eee9ec/index.js#L487-L499) 
```ts
 async function getMultipartFile (options) {
    const parts = this[kMultipartHandler](options)
    let part
    while ((part = await parts()) != null) {
      if (part.file) {
        // part.file.truncated is true when a configured file size limit is reached
        if (part.file.truncated && throwFileSizeLimit) {
          throw new RequestFileTooLargeError()
        }
        return part
      }
    }
  }
```
As you can see, this only loads files, never loading field - thus 
```
const cacheTime = data.fields.cacheControl?.value
```
will never have a value.

The solution would be to use `parts()` API that returns an AysncIterable of Multipart and process that or give the value as a header to make life easier.

**Comment by inian** [comment]
Thanks for digging into this @horizon0708. Looks like the non file values need to be sent _before_ the file fields as mentioned [here](https://github.com/fastify/fastify-multipart/issues/146#issuecomment-681874488) and it is the intended behaviour. Fixing in supabase-js

**Comment by inian** [comment]
Can you try with supabase-js 1.11.14 James?

**Comment by horizon0708** [comment]
I think it's working! Thanks for a quick update, looks like I was completely off the mark with https://github.com/supabase/storage-api/issues/18#issuecomment-843717897😅
![image](https://user-images.githubusercontent.com/22534929/118761712-02cf2800-b8c9-11eb-82ab-b430a1cf1022.png)

**Comment by inian** [comment]
I thought it might have to do with something with Fastify as well till I realised that it was working with when I sent the request via Postman but not via supabase-js
